gnrc_netif: highest source address scope wins selection

miri64 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
Rule 2 of the source address algorithm outlined in [RFC6724] states the possible source addresses must also be compared among each other:

>     Rule 2: Prefer appropriate scope.
>     If Scope(SA) < Scope(SB): If Scope(SA) < Scope(D), then prefer SB and
>     otherwise prefer SA.  Similarly, if Scope(SB) < Scope(SA): If
>     Scope(SB) < Scope(D), then prefer SA and otherwise prefer SB.

Our current implementation doesn't do that. It just checks if the scope of a possible source is lesser than the scope of the destination (which involves the second "If" in the rule).

This fix grants points according to the scope of an address. If the scope matches, they get the highest points, ensuring that the selected source will always be reachable from the destination.

[RFC6724]: https://tools.ietf.org/html/rfc6724

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
Pinging between different and same scopes should still work (or even work for the first time).
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
Follow-up to #12404 and  #12408.
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

tcschmidt commented:
@cgundogan can you test against the RFC spec?

miri64 commented:
@cgundogan @tcschmidt please be aware that without #12408 the result of the source address selection gets skewed to the first address assigned to the interface in any case, so the scope contention might not even come through.

fjmolinas commented:
@miri64 this one needs rebasing now.

miri64 commented:
Rebased to current master. Please make sure to test this also wrt site-local addresses.

benpicco commented:
How would the testing procedure look like / how did you test this?

miri64 commented:
> How would the testing procedure look like / how did you test this?

I tested it like I described in the testing procedure, I think (sorry I don't remember exactly). I think for a more in-depth test, one needs to find a set of addresses, where all the other rules result in equal ranking but not the scope and use one of them as destination address and the rest is assigned to the interface as potential source addresses. The largest scope then should win. There can only be three scopes: global, site-local and link-local from largest to smallest.

miri64 commented:
> that being said I'm having trouble to understand how I should test this, I understand the test cases but to be completely honest I dont understand how to test scopes regarding multicast address. Should I just add them to the group and ping that group?

There is no involvement of multicast here (unless of course you want to use it as a destination address, but you don't have to). The scope values from multicast addresses are just used because they happen to provide a standadized numerical order to scopes (I believe; just re-using what was there from @OlegHahm's implementation).

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1067,11 +1064,11 @@ static ipv6_addr_t *_src_addr_selection(gnrc_netif_t *netif,
         uint8_t candidate_scope = _get_scope(ptr);
```
:woman_shrugging: It doesn't even do that in the first place. There is just this comment saying "both link local", but it does not say what "both" is referring to and if this is an assumption or something else. This comment goes back to the [original implementation of this function](https://github.com/RIOT-OS/RIOT/pull/3561/files#diff-db9a1a910efb231b7ee7971f29055350R603) and even there it doesn't to anything with link local etc. It is not in the code I touched so I deem it unrelated, but if you want, I can add a piggy-back commit to remove it.

fjmolinas replied:
> It is not in the code I touched so I deem it unrelated, but if you want, I can add a piggy-back commit to remove it.

Yes it is unrelated but since if was under rule 2, It got me confused. You can add the commit or not, as you said it is unrelated.

fjmolinas commented:
> There is no involvement of multicast here (unless of course you want to use it as a destination address, but you don't have to). The scope values from multicast addresses are just used because they happen to provide a standadized numerical order to scopes (I believe; just re-using what was there from @OlegHahm's implementation).

If there is no involvement of multicast how can the scopes be different? From what I understand unicast can have either gobal or local scope. So there would be no case for two candidates to have smaller scope but not equal right? either they both have smaller and equal or one have the same scope.

fjmolinas commented:
Ok so I have two nodes with link-local and global scope:

<details><summary>node1 ifconfig</summary>

```
Iface  6  HWaddr: 11:15  Channel: 26  Page: 0  NID: 0x23
          Long HWaddr: 06:BC:FB:65:10:6B:11:15 
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::4bc:fb65:106b:1115  scope: link  VAL
          inet6 addr: 2001:6d6f:6c69:6e61:4bc:fb65:106b:1115  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff6b:1115
          

```
</details>
<details><summary>node2 ifconfig</summary>

```
Iface  7  HWaddr: 0F:36  Channel: 26  Page: 0  NID: 0x23
          Long HWaddr: 15:11:6B:10:65:FA:8F:36 
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::1711:6b10:65fa:8f36  scope: link  VAL
          inet6 addr: 2001:6d6f:6c69:6e61:1711:6b10:65fa:8f36  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:fffa:8f36
          inet6 group: ff02::1a

```
</details>

```
ping6 2001:6d6f:6c69:6e61:4bc:fb65:106b:1115 
2020-01-10 17:48:46,100 #  ping6 2001:6d6f:6c69:6e61:4bc:fb65:106b:1115
2020-01-10 17:48:46,102 # finding the best match within the source address candidates
2020-01-10 17:48:46,104 # Checking address: fe80::1711:6b10:65fa:8f36
2020-01-10 17:48:46,105 # winner for rule 2 (smaller scope) found
2020-01-10 17:48:46,107 # winner for rule 3 found
2020-01-10 17:48:46,108 # Checking address: 2001:6d6f:6c69:6e61:1711:6b10:65fa:8f36
2020-01-10 17:48:46,109 # winner for rule 2 (same scope) found
2020-01-10 17:48:46,110 # winner for rule 3 found
2020-01-10 17:48:46,111 # Winner is: 2001:6d6f:6c69:6e61:1711:6b10:65fa:8f36

```

```
ping6 fe80::4bc:fb65:106b:1115 
2020-01-10 17:49:18,194 # ping6 fe80::4bc:fb65:106b:1115
2020-01-10 17:49:18,197 # finding the best match within the source address candidates
2020-01-10 17:49:18,199 # Checking address: fe80::1711:6b10:65fa:8f36
2020-01-10 17:49:18,200 # winner for rule 2 (same scope) found
2020-01-10 17:49:18,201 # winner for rule 3 found
2020-01-10 17:49:18,203 # Checking address: 2001:6d6f:6c69:6e61:1711:6b10:65fa:8f36
2020-01-10 17:49:18,204 # winner for rule 3 found
2020-01-10 17:49:18,206 # Winner is: fe80::1711:6b10:65fa:8f36
2020-01-10 17:49:18,211 # 12 bytes from fe80::4bc:fb65:106b:1115: icmp_seq=0 ttl=64 rssi=-58 dBm time=13.421 ms

```

That works correctly.

miri64 commented:
> If there is no involvement of multicast how can the scopes be different? From what I understand unicast can have either gobal or local scope. So there would be no case for two candidates to have smaller scope but not equal right? either they both have smaller and equal or one have the same scope.

There *are* the [deprecated](https://tools.ietf.org/html/rfc3879) [site-local address](https://tools.ietf.org/html/rfc4291#section-2.5.7), which would provide another level of scope.

miri64 commented:
There are still problems with wide-scoped multicast addresses so I would not use them for testing. See https://github.com/RIOT-OS/RIOT/issues/4527 and https://github.com/RIOT-OS/RIOT/issues/5230.

fjmolinas commented:
In RFC 4291: `Site-Local IPv6 Unicast Addresses`

```
The special behavior of this prefix defined in [RFC3513] must no
   longer be supported in new implementations (i.e., new implementations
   must treat this prefix as Global Unicast).
```

So I guess alltough good, it doesn't need to be supported?

miri64 commented:
For multicast AFAIK site local is not deprecated and it is the same check (in a function I don't touch in this PR), so no sense in removing it here.

fjmolinas commented:
Thanks @miri64, sorry for asking so many questions. I'm got got quite confused between the RFC's. I changed my setup, two nodes with these address:


node1:
```
2020-01-10 18:20:21,972 #           inet6 addr: fe80::7b65:3458:ecde:6b82  scope: link  VAL
2020-01-10 18:20:21,978 #           inet6 addr: fd01::7b65:3458:ecde:6b82  scope: global  VAL
```

node2:
```
2020-01-10 18:19:11,656 #           inet6 addr: fd01::1711:6b10:65fa:8f36  scope: global  VAL
2020-01-10 18:19:11,658 #           inet6 addr: fec0::1711:6b10:65fa:8f36  scope: site  VAL
```

From node2 to node1 link local, site local wins. (Second part of rule 2 working :) )

```
ping6 fe80::7b65:3458:ecde:6b82
2020-01-10 18:21:42,352 # ping6 fe80::7b65:3458:ecde:6b82
2020-01-10 18:21:42,354 # finding the best match within the source address candidates
2020-01-10 18:21:42,356 # Checking address: fd01::1711:6b10:65fa:8f36
2020-01-10 18:21:42,357 # winner for rule 3 found
2020-01-10 18:21:42,359 # Checking address: fec0::1711:6b10:65fa:8f36
2020-01-10 18:21:42,360 # winner for rule 3 found
2020-01-10 18:21:42,362 # Winner is: fec0::1711:6b10:65fa:8f36
2020-01-10 18:21:43,360 # finding the best match within the source address candidates
2020-01-10 18:21:43,362 # Checking address: fd01::1711:6b10:65fa:8f36
2020-01-10 18:21:43,363 # winner for rule 3 found
2020-01-10 18:21:43,364 # Checking address: fec0::1711:6b10:65fa:8f36
2020-01-10 18:21:43,365 # winner for rule 3 found
2020-01-10 18:21:43,366 # Winner is: fec0::1711:6b10:65fa:8f36
2020-01-10 18:21:44,353 # finding the best match within the source address candidates
2020-01-10 18:21:44,355 # Checking address: fd01::1711:6b10:65fa:8f36
2020-01-10 18:21:44,356 # winner for rule 3 found
2020-01-10 18:21:44,358 # Checking address: fec0::1711:6b10:65fa:8f36
2020-01-10 18:21:44,359 # winner for rule 3 found
2020-01-10 18:21:44,360 # Winner is: fec0::1711:6b10:65fa:8f36
```

From node2 to node1 global, global wins.

```
> ping6 fd01::7b65:3458:ecde:6b82 
2020-01-10 18:22:17,726 #  ping6 fd01::7b65:3458:ecde:6b82
2020-01-10 18:22:17,728 # finding the best match within the source address candidates
2020-01-10 18:22:17,730 # Checking address: fd01::1711:6b10:65fa:8f36
2020-01-10 18:22:17,732 # winner for rule 2 (same scope) found
2020-01-10 18:22:17,732 # winner for rule 3 found
2020-01-10 18:22:17,741 # Checking address: fec0::1711:6b10:65fa:8f36
2020-01-10 18:22:17,743 # winner for rule 2 (smaller scope) found
2020-01-10 18:22:17,744 # winner for rule 3 found
2020-01-10 18:22:17,745 # Winner is: fd01::1711:6b10:65fa:8f36
2020-01-10 18:22:18,734 # finding the best match within the source address candidates
2020-01-10 18:22:18,736 # Checking address: fd01::1711:6b10:65fa:8f36
2020-01-10 18:22:18,738 # winner for rule 2 (same scope) found
2020-01-10 18:22:18,739 # winner for rule 3 found
2020-01-10 18:22:18,741 # Checking address: fec0::1711:6b10:65fa:8f36
2020-01-10 18:22:18,742 # winner for rule 2 (smaller scope) found
2020-01-10 18:22:18,743 # winner for rule 3 found
2020-01-10 18:22:18,745 # Winner is: fd01::1711:6b10:65fa:8f36
2020-01-10 18:22:19,727 # finding the best match within the source address candidates
2020-01-10 18:22:19,728 # Checking address: fd01::1711:6b10:65fa:8f36
2020-01-10 18:22:19,730 # winner for rule 2 (same scope) found
2020-01-10 18:22:19,731 # winner for rule 3 found
2020-01-10 18:22:19,741 # Checking address: fec0::1711:6b10:65fa:8f36
2020-01-10 18:22:19,743 # winner for rule 2 (smaller scope) found
2020-01-10 18:22:19,744 # winner for rule 3 found
2020-01-10 18:22:19,745 # Winner is: fd01::1711:6b10:65fa:8f36
```

fjmolinas commented:
@miri64 can you confirm that might output makes sense :D ?

fjmolinas commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1063,15 +1060,14 @@ static ipv6_addr_t *_src_addr_selection(gnrc_netif_t *netif,
-            winner_set[i] += RULE_2A_PTS;
+            winner_set[i] += RULE_2_PTS;
         }
         else if (candidate_scope < dst_scope) {
```
`If Scope(SA) < Scope(SB): If Scope(SA) < Scope(D), then prefer SB and`

To me this means prefer scope closed to D but that are greater than D, in that case you should get points if you are bugger than destination

```suggestion
        else if (candidate_scope > dst_scope) {
```

miri64 replied:
I think you are right on that.

fjmolinas commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1063,15 +1060,14 @@ static ipv6_addr_t *_src_addr_selection(gnrc_netif_t *netif,
         else if (candidate_scope < dst_scope) {
             DEBUG("winner for rule 2 (smaller scope) found\n");
-            winner_set[i] += RULE_2B_PTS;
+            winner_set[i] += candidate_scope;
```
```suggestion
            winner_set[i] += candidate_scope - dst_scope;
```

miri64 replied:
Doesn't this invert the scoring for scopes. With that the lesser scope gets more precedence.

fjmolinas replied:
And that is how I understand it should be. I understand `Rule 2` as choose the scope closest to D that is still greater than D.

```
Rule 2: Prefer appropriate scope.
If Scope(SA) < Scope(SB): If Scope(SA) < Scope(D), then prefer SB and
otherwise prefer SA.
```

The above states the case:

- `Scope(SA) < Scope(D) < Scope(SB)` choose SB
- `Scope(D) < Scope(SA) < Scope(SB)` choose SA

miri64 replied:
Yes, but with you suggestion it becames

```
Rule 2: Prefer appropriate scope.
If Scope(SA) > Scope(SB): If Scope(SA) < Scope(D), then prefer SB and
otherwise prefer SA.
```

miri64 replied:
(as opposed to `Scope(SA) < Scope(SB)`)

miri64 replied:
>     * `Scope(SA) < Scope(D) < Scope(SB)` choose SB

Also according to rule 2 "`Scope(SA) < Scope(SB) < Scope(D)` choose SB" btw.

fjmolinas replied:
`Also according to rule 2 "Scope(SA) < Scope(SB) < Scope(D) choose SB" btw.`

yes you are right, sorry I got confused by reading doc on `dest` and `source` address selection.

But Lesser scope should still be prefered in any other case than `Scope(SA) < Scope(SB) < Scope(D)` right?

miri64 replied:
Okay, let's follow this through. Let's say there are three addresses `SA`, `SB`, and `SC` assigned to the interface and the scopes are as follows:

```
Scope(SA) < Scope(D) < Scope(SB) < Scope(SC)
```

(i know this is not possible with unicast addresses, but let's go through this anyway, to make sure we got this right)

Accoding to rule 2 `SB` should be selected.

If we go through it as proposed by you then for `SA` this line is skipped as `candidate_scope < dst_scope`, for `SB` `winner_set[pos(SB)]` is incremented by `RULE_2_PTS - Scope(SB)`, selecting `SB` for now. For `SC` `winner_set[pos(SC)]` is incremented by `RULE_2_PTS - Scope(SC)` which does not overtake `winner_set[pos(SB)]` as `(RULE_2_PTS - Scope(SB)) > RULE_2_PTS - Scope(SC)`.

I think you are right.

fjmolinas replied:
@miri64 doesn't the rule mean (easier to map with code hahaha):

```
if (A < B) {
    if( A < D) {
        choose B
    }
}
else {
    chose A
}
```

miri64 commented:
Addressed comments. I now have a knot in my brain :sweat_smile:. I also added as a separate commit the proper reading of the scope of multicast addresses.

fjmolinas commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -1067,7 +1072,16 @@ static ipv6_addr_t *_src_addr_selection(gnrc_netif_t *netif,
         }
         else if (candidate_scope > dst_scope) {
             DEBUG("winner for rule 2 (larger scope) found\n");
-            winner_set[i] += candidate_scope;
```
The only issue is the case `source(SA) < source(SB) < source(D)`, I find sources that say that in that case they both have the same preference and others that sat `SB` should be preferred. The `RFC` makes me think it should be `SA`, but with the `if (candidate_scope > dst_scope)` they will both get the same preference.

```
if (A < B) {
    if( A < D) {
        choose B
    } else {
        choose A
    }
}
else {
    choose A
}
```


BTW: its been a while since something so simple has given me sucha  headache ahahaha, just 2 ifs

miri64 replied:
> BTW: its been a while since something so simple has given me sucha headache ahahaha, just 2 ifs

Yeah... but the wording in the RFC could be a little bit more trivial IMHO so no wonder it gives a headache :sweat_smile:

miri64 replied:
Actually I agree with you: the RFC is pretty clear on the case you talked about: It should be `SB`: `if Scope(SA) < Scope(SB)` (true in our case) then `If Scope(SA) < Scope(D),` (also true in our case) pick `then prefer SB […]`.

miri64 replied:
(so we need an else case where `SB` is picked if `Scope(SA) < Scope(SB) < Scope(D)`, but `SC`, if `Scope(SA) < Scope(SB) < Scope(D) < Scope(SC)`... I think that makes the scoring a bit more complicated ^^"

miri64 replied:
@fjmolinas I think my latest proposal should make the best selection.

miri64 commented on sys/net/gnrc/netif/gnrc_netif.c:
```diff
@@ -892,7 +892,12 @@ static int _match_to_idx(const gnrc_netif_t *netif,
 static uint8_t _get_scope(const ipv6_addr_t *addr)
 {
-    if (ipv6_addr_is_link_local(addr)) {
+    if (ipv6_addr_is_multicast(addr)) {
```
I wanted a fixup. Mainly, because it is a bit beyond the original intention of this PR and I wanted to here your opinion first. I can make it its own commit later-on still.

fjmolinas replied:
I think it should be a separate commit. I don't mind including it, its needed for multicast but in a different commit.

fjmolinas commented:
- Scope(D) < Scope(SA) < Scope(SB): **Scope(SA)** wins

<details>

```
2020-01-13 17:15:17,285 #           inet6 addr: fd01::1711:6b10:65fb:bc06  scope: global  VAL
2020-01-13 17:15:17,286 #           inet6 addr: fec0::1711:6b10:65fb:bc06  scope: site  VAL
```

```
2020-01-13 17:15:35,620 #  ping6 fe80::1711:6b10:65fb:bc06
2020-01-13 17:15:35,621 # finding the best match within the source address candidates
2020-01-13 17:15:35,621 # Checking address: fd01::1711:6b10:65fb:bc06
2020-01-13 17:15:35,622 # winner for rule 2 (larger scope) found
2020-01-13 17:15:35,622 # winner for rule 3 found
2020-01-13 17:15:35,623 # Checking address: fec0::1711:6b10:65fb:bc06
2020-01-13 17:15:35,623 # winner for rule 2 (larger scope) found
2020-01-13 17:15:35,623 # winner for rule 3 found
2020-01-13 17:15:35,623 # Winner is: fec0::1711:6b10:65fb:bc06
```
</details>

- Scope(SA) < Scope(SB) <  Scope(D): **Scope(SB)** wins

<details>

```
2020-01-13 17:12:41,662 #           inet6 addr: fe80::1711:6b10:65fb:bc06  scope: link  VAL
2020-01-13 17:12:41,662 #           inet6 addr: fec0::1711:6b10:65fb:bc06  scope: site  VAL
```

```
2020-01-13 17:14:12,326 # ping6 fd01::7b65:3458:ecde:6b82
2020-01-13 17:14:12,327 # finding the best match within the source address candidates
2020-01-13 17:14:12,327 # Checking address: fe80::1711:6b10:65fb:bc06
2020-01-13 17:14:12,328 # winner for rule 2 (smaller scope) found
2020-01-13 17:14:12,328 # winner for rule 3 found
2020-01-13 17:14:12,329 # Checking address: fec0::1711:6b10:65fb:bc06
2020-01-13 17:14:12,330 # winner for rule 2 (smaller scope) found
2020-01-13 17:14:12,330 # winner for rule 3 found
2020-01-13 17:14:12,331 # Winner is: fec0::1711:6b10:65fb:bc06
```
</details>

- Scope(SA) < Scope(D) < Scope(SB): **Scope(SB)** wins

<details>

```
2020-01-13 17:16:27,247 #           inet6 addr: fd01::1711:6b10:65fb:bc06  scope: global  VAL
2020-01-13 17:16:27,247 #           inet6 addr: fe80::1711:6b10:65fb:bc06  scope: link  VAL
```
```
2020-01-13 17:16:32,346 # ping6 fec0::1711:6b10:65fb:bc06
2020-01-13 17:16:32,347 # finding the best match within the source address candidates
2020-01-13 17:16:32,347 # Checking address: fd01::1711:6b10:65fb:bc06
2020-01-13 17:16:32,348 # winner for rule 2 (larger scope) found
2020-01-13 17:16:32,348 # winner for rule 3 found
2020-01-13 17:16:32,348 # Checking address: fe80::1711:6b10:65fb:bc06
2020-01-13 17:16:32,349 # winner for rule 2 (smaller scope) found
2020-01-13 17:16:32,349 # winner for rule 3 found
2020-01-13 17:16:32,349 # Winner is: fd01::1711:6b10:65fb:bc06
```
</details>


- Scope(SA) = Scope(D) < Scope(SB): **Scope(SA)** wins

<details>

```
2020-01-13 17:17:02,592 #           inet6 addr: fd01::1711:6b10:65fb:bc06  scope: global  VAL
2020-01-13 17:17:02,596 #           inet6 addr: fe80::1711:6b10:65fb:bc06  scope: link  VAL
```
```
ping6 fe80::1711:6b10:65fb:bc06 
2020-01-13 17:17:08,035 #  ping6 fe80::1711:6b10:65fb:bc06
2020-01-13 17:17:08,037 # finding the best match within the source address candidates
2020-01-13 17:17:08,038 # Checking address: fd01::1711:6b10:65fb:bc06
2020-01-13 17:17:08,039 # winner for rule 2 (larger scope) found
2020-01-13 17:17:08,039 # winner for rule 3 found
2020-01-13 17:17:08,051 # Checking address: fe80::1711:6b10:65fb:bc06
2020-01-13 17:17:08,051 # Ease one - rule 1
```
</details>

- Scope(SA) < Scope(D) = Scope(SB): **Scope(SB)** wins

<details>

```
2020-01-13 17:17:02,592 #           inet6 addr: fd01::1711:6b10:65fb:bc06  scope: global  VAL
2020-01-13 17:17:02,596 #           inet6 addr: fe80::1711:6b10:65fb:bc06  scope: link  VAL
```
```
2020-01-13 17:19:02,637 #  ping6 fd01::1711:6b10:65fb:bc06
2020-01-13 17:19:02,638 # finding the best match within the source address candidates
2020-01-13 17:19:02,639 # Checking address: fd01::1711:6b10:65fb:bc06
2020-01-13 17:19:02,639 # Ease one - rule 1
```
</details>

-------------------------------------------------------------------------------------------------------------------------------

Multicast test, just to verify its returned correctly.

- Scope(SA) < Scope(D) < Scope(SB): **Scope(SB)** wins

<details>

```
2020-01-13 17:17:02,592 #           inet6 addr: fd01::1711:6b10:65fb:bc06  scope: global  VAL
2020-01-13 17:17:02,596 #           inet6 addr: fe80::1711:6b10:65fb:bc06  scope: link  VAL
```
```
2020-01-13 17:25:15,148 #  ping6 ff04::1
2020-01-13 17:25:15,149 # finding the best match within the source address candidates
2020-01-13 17:25:15,149 # Checking address: fd01::1711:6b10:65fb:bc06
2020-01-13 17:25:15,149 # winner for rule 2 (larger scope) found
2020-01-13 17:25:15,150 # winner for rule 3 found
2020-01-13 17:25:15,150 # Checking address: fe80::1711:6b10:65fb:bc06
2020-01-13 17:25:15,150 # winner for rule 2 (smaller scope) found
2020-01-13 17:25:15,151 # winner for rule 3 found
2020-01-13 17:25:15,151 # Winner is: fd01::1711:6b10:65fb:bc06
```
</details>

- Scope(SA) = Scope(D) < Scope(SB): **Scope(SA)** wins

<details>

```
2020-01-13 17:17:02,592 #           inet6 addr: fd01::1711:6b10:65fb:bc06  scope: global  VAL
2020-01-13 17:17:02,596 #           inet6 addr: fe80::1711:6b10:65fb:bc06  scope: link  VAL
```
```
2020-01-13 17:25:59,306 #  ping6 ff02::1
2020-01-13 17:25:59,309 # finding the best match within the source address candidates
2020-01-13 17:25:59,311 # Checking address: fd01::1711:6b10:65fb:bc06
2020-01-13 17:25:59,322 # winner for rule 2 (larger scope) found
2020-01-13 17:25:59,323 # winner for rule 3 found
2020-01-13 17:25:59,324 # Checking address: fe80::1711:6b10:65fb:bc06
2020-01-13 17:25:59,325 # winner for rule 2 (same scope) found
2020-01-13 17:25:59,325 # winner for rule 3 found
2020-01-13 17:25:59,326 # Winner is: fe80::1711:6b10:65fb:bc06
```
</details>

fjmolinas commented:
@miri64 Now its perfect! Please squash and split 03da7f1cb05f93ee1ebe148840502de9cfe2c085 in to its own commit.

miri64 commented:
Squashed.

miri64 commented:
:tada: Thanks for all testing and helping with the logic foobar of this PR @fjmolinas!
